File: backend/app/models/scan.py
from pydantic import BaseModel, UUID4
from datetime import datetime
from typing import Optional


class ScanStatus(BaseModel):
    id: UUID4
    triggered_at: datetime
    completed_at: Optional[datetime] = None
    mode: str
    status: str
    universe_size: Optional[int] = None
    passed_filter: Optional[int] = None
    signals_found: Optional[int] = None
    duration_sec: Optional[float] = None
    error_message: Optional[str] = None


class TriggerScanRequest(BaseModel):
    mode: str = "daily"          # daily | full | fno | custom
    universe: list[str] = []     # empty = use default universe for mode


class TriggerScanResponse(BaseModel):
    scan_id: str
    status: str = "pending"

File: apps/web/src/types/scan.ts
import type { SignalCategory, SignalSide } from "./signal";

export interface ScanResult {
  id: string;
  mode: string;
  status: "pending" | "running" | "done" | "failed";
  triggered_at: string;
  completed_at: string | null;
  signals_found: number | null;
  passed_filter: number | null;
  universe_size: number | null;
  duration_sec: number | null;
  error_message: string | null;
}

/** Partial signal shape received over WebSocket before DB persists full record. */
export interface StreamingSignal {
  symbol: string;
  category: SignalCategory;
  side: SignalSide | null;
  signal_timeframe: string | null;
  entry: number | null;
  stop_loss: number | null;
  t1: number | null;
  t2: number | null;
  t3: number | null;
  rr_t1: number | null;
  rr_t2: number | null;
  gate_strength: number | null;
  confidence: number | null;
  rank_score: number | null;
  htf_confirmed: boolean | null;
}
